### PR TITLE
Toggle floating badge visibility onscroll

### PR DESCRIPTION
**Pull request by kosdiamantis**
Specs:

```
Badge
  floating
    show/hide onscroll
      when hide_on_scroll set to true
        on small screen
          ✔ hides floating badge on scroll down
          ✔ shows floating badge on scroll up
          when scoll horizontally
            ✔ does not change badge visibility
        on big screen
          ✔ does not change badge visibility
      when hide_on_scroll set to false
        on small screen
          ✔ does not change badge visibility
        on big screen
          ✔ does not change badge visibility
```

### CODE REVIEW

File: src/plugins/badge.coffee [hunk 1]
```diff
@@ -608,4 +610,45 @@ class Badge
...
+
+    past_top || past_bottom
+
+  _smallDevice: -> @_vieport().width <= 768
+
+  _vieport: ->
+    e = window.parent
+    a = 'inner'
+
+    if !('innerWidth' of e)
```
Comment (harrisred) [inline src/plugins/badge.coffee]: Could use `unless`, and `hasOwnProperty` seems a better fit as we don't want to check for an inherited property.
`unless e.hasOwnProperty('innerWidth')`

----------
And I suggest a different approach, which I think is simpler and more readable:
```
_vieport: ->
    if window.parent.hasOwnProperty('innerWidth')
      width: window.parent.innerWidth
      heigth: window.parent.innerWidth
    else
      width: (@parent_doc.documentElement or @parent_doc.body).clientWidth
      height: (@parent_doc.documentElement or @parent_doc.body).clientHeight
```

File: src/plugins/badge.coffee [hunk 2]
```diff
@@ -608,4 +610,45 @@ class Badge
...
+      return if Math.abs(scroll_diff) <= 5
+
+      badge.style.display = if scroll_diff > 0 then 'none' else 'block'
+
+      past_scroll = current_scroll
+
+  _outOfBound: (current_scroll) ->
+    past_top = current_scroll < 0
+
+    document_height = Math.max @parent_doc.body.scrollHeight, @parent_doc.documentElement.scrollHeight,
```
Comment (harrisred) [inline src/plugins/badge.coffee]: Could extracted it to a separate private method

File: src/plugins/badge.coffee [hunk 3]
```diff
@@ -608,4 +610,45 @@ class Badge
...
+      return if @_outOfBound(current_scroll) # Ignore OSX bounce
+
+      scroll_diff = current_scroll - past_scroll
+      return if Math.abs(scroll_diff) <= 5
+
+      badge.style.display = if scroll_diff > 0 then 'none' else 'block'
+
+      past_scroll = current_scroll
+
+  _outOfBound: (current_scroll) ->
```
Comment (harrisred) [inline src/plugins/badge.coffee]: Alternative that uses early return, without making all the calculations for the bottom bounce:

```
  _outOfBound: (current_scroll) ->
    return true if current_scroll < 0 #top bounce

    current_scroll + @_vieport().height > @_document_height() #bottom bounce

  _document_height: ->
    Math.max @parent_doc.body.scrollHeight, @parent_doc.documentElement.scrollHeight,
             @parent_doc.body.offsetHeight, @parent_doc.documentElement.offsetHeight,
             @parent_doc.body.clientHeight, @parent_doc.documentElement.clientHeight
```

Comment (kosdiamantis) [inline src/plugins/badge.coffee]: 👌

File: src/plugins/badge.coffee (hunk 2 above)
Comment (kosdiamantis) [inline src/plugins/badge.coffee]: 👌

File: src/plugins/badge.coffee (hunk 1 above)
Comment (kosdiamantis) [inline src/plugins/badge.coffee]: 👌

File: src/plugins/badge.coffee [hunk 4]
```diff
@@ -624,7 +645,12 @@ class Badge
       scroll_diff = current_scroll - past_scroll
       return if Math.abs(scroll_diff) <= 5
 
-      badge.style.display = if scroll_diff > 0 then 'none' else 'block'
+      if scroll_diff > 0
```
Comment (harrisred) [inline src/plugins/badge.coffee]: [super minor] Could add a comment `# scroll down`

Comment (kosdiamantis) [inline src/plugins/badge.coffee]: 👌

**Pull request merged**